src/relay/backend/vm/manifest_lifetimes.cc

```diff
@@ -0,0 +1,611 @@
+
+  /*!
+   * \brief Analyze the input \p cfg (using info from \p use_def).
+   *
+   * \param cfg The input control flow graph.
+   * \param use_def Use-def analysis of \p cfg.
+   * \return LivenessAnalysis
+   */
+  static LivenessAnalysis Analyze(const ControlFlowGraph& cfg, const UseDefAnalysis& use_def) {
+    LivenessAnalysis a;
+    std::list<CFG::NodePtr> worklist;
+
+    // Initialize worklist to post-order traversal for quick convergence.
+    worklist.insert(worklist.end(), cfg.reverse_post_order.rbegin(), cfg.reverse_post_order.rend());
+
+    // See https://lambda.uta.edu/cse5317/notes/node40.html for an overview of the algorithm.
+    auto visitor = [&](const CFG::NodePtr n) {
+      VarSet old_in_n = a.live_in[n];
+      VarSet old_out_n = a.live_out[n];
+
+      a.live_in[n] = use_def.use.at(n);
+      for (const Var& v : a.live_out[n]) {
+        if (!v.same_as(use_def.def.at(n))) {
+          a.live_in[n].insert(v);
+        }
+      }
+
+      a.live_out[n] = VarSet();
+      for (const CFG::NodePtr& s : n->GetSucc()) {
+        a.live_out[n].insert(a.live_in[s].begin(), a.live_in[s].end());
+      }
+
+      if (SetEqual(old_in_n, a.live_in[n]) && SetEqual(old_out_n, a.live_out[n])) {
+        // No need to update the worklist.
+      } else {
+        // Add predecessor nodes back to worklist (no need to add successors, since each node's
+        // in/out sets are not dependent on its predecessors).
+        for (const CFG::NodePtr& p : n->GetPred()) {
+          worklist.push_back(p);
+        }
+      }
+    };
+
+    while (!worklist.empty()) {
+      const CFG::NodePtr n = worklist.front();
+      worklist.pop_front();
+      visitor(n);
+    }
+
+    return a;
+  }
+};
+
+/*!
+ * \brief Helper class to insert kills using liveness information.
+ */
+class KillInserter : public ExprMutator {
+ public:
+  KillInserter(const ControlFlowGraph* cfg, const LivenessAnalysis* lva) : cfg_(cfg), lva_(lva) {}
+
+  // Limitations
+  // -----------
+  // 1. For simplicity, we only insert kills when visiting Let bindings, and always emit the kill as
+  // a single subsequent binding. This is slightly inaccurate; for example, if the condition of an
+  // If is dead after the test, we can immediately kill the condition in each branch:
+  //   let %x = if (%dead_cond) {
+  //     let %_0 = memory.kill(%dead_cond);
+  //     ...
+  //   } else {
+  //     let %_1 = memory.kill(%dead_cond);
+  //     ...
+  //   }
+  // as opposed to:
+  //   let %x = if (%dead_cond) ...
+  //   let %_0 = memory.kill(%dead_cond);
+  //
+  // 2. Killed variables are calculated as live in - live out, which misses variables that are
+  // actually dead but not in live in. Examples include: when the last use of a var is the result
+  // expr of an If branch; when bound vars (i.e. function inputs, pattern matched vars, dead
+  // bindings) are never used.
+  //
+  // 3. When the result expr of an If branch is a variable, and this expr is the last use of the
```

Comment:
Shoot, I missed that one. Could push the 'return kills' onto an internal stack and pop them after each join point, but agree a TODO is find.